### PR TITLE
[CI] [MLA] Deterministic single-split decode option for reproducible serving (fixes #4364)

### LINKED ISSUE #4364: [MLA] qh16 fp8 sparse decode run-to-run nondeterministic (split-K reduce) — ±1-2 NIAH needle
## Summary

With the gqa64→qh16 fold in place (see the qh64 `page_size=1` GPU-fault issue #4363),
GLM-5.1-FP8 DSA long-context decode is **run-to-run nondeterministic** on repeated,
identical Needle-In-A-Haystack (NIAH) requests at temperature 0: ±1–2 retrieved
needles at mid-range context (20k–35k), while the extremes (8k, 96k) are stable.
This points at the split-K reduce accumulation order in the qh16 fp8 sparse decode path.

## Evidence (found/10, 3 identical trials, temp=0)

1P1D EP8 (DP=8/TP=1 per role):

```
8k   -> 10 / 10 / 10   (deterministic)
20k  ->  9 /  9 /  8
35k  ->  9 /  9 / 10
96k  -> 10 / 10 / 10
```

2P2D EP16:

```
8k   -> 10 / 10 / 10   (deterministic)
20k  ->  8 / 10 /  8
35k  -> 10 / 10 / 10
96k  -> 10 / 10 / 10
```

No crash; correctness is otherwise high. The variance is stable across seeds/reruns
and consistent with a nondeterministic split-K reduce rather than a modeling bug.

## Question / ask

Is the qh16 fp8 sparse MLA decode split-K reduce deterministic-orderable for temp=0
serving (fixed split count / stable reduction order)? For customer serving that needs
reproducible outputs, a deterministic reduce option on this path would help.

Env: ROCm 7.2.3, MI300X (gfx942), AITER v0.1.18, vLLM 0.25.1 + PR#47766, GLM-5.1-FP8.

Related: #4363 (the gqa64→qh16 fold that this nondeterminism rides on).

### PR DESCRIPTION
**Pull request by MohitAMD**
## Motivation

Closes #4364.

The MLA decode **split-KV reduce is fp-non-associative**, so a request's decode
output depends on *how many* KV splits it is given. The per-request split count is
drawn from the **global** budget `min(cu_num, max_split_per_batch * batch_size)`
(see `csrc/kernels/mla/metadata/v1_2_device.cuh`), which **varies with batch
composition**. Two identical requests at temperature 0 can therefore land on
different split counts on different steps and produce slightly different logits.

On GLM-5.1-FP8 (DSA, qh16 fp8 decode) this surfaces as run-to-run
nondeterminism — ±1–2 retrieved NIAH needles at mid context (20k–35k), while the
extremes (8k, 96k) are stable (#4364).

## Reproducer

`op_tests/test_mla_qh16_decode_determinism.py` (added here) isolates the mechanism
on a single MI300X (gfx942), nhead=16 fp8/fp8 persistent decode, batch=1, ctx=8192:

```
[det] split=1 run-to-run  max|d| = 0.000e+00   (single-split reduce is a no-op)
[det] split=8 run-to-run  max|d| = 0.000e+00   (kernel is deterministic for FIXED metadata)
[det] split=1 vs split=8 max|d| = 2.441e-03   (the split COUNT changes the numerics)
```

So the kernel is bit-stable for a fixed split count, but the result **depends on
the split count** — and production picks that count from the batch-dependent
budget, which is why identical requests drift run-to-run.

## Workaround (this PR)

Opt-in env knobs, honored in `get_mla_metadata_v1`:

- `AITER_MLA_DECODE_DETERMINISTIC=1` — force a **single split per request**
  (shorthand for the cap below = 1).
- `AITER_MLA_DECODE_MAX_SPLIT_PER_BATCH=<n>` — pin the per-batch split cap to a
  fixed `n` (≥1), independent of batch size.

Forcing a single split makes the reduce a no-op, so a request's output no longer
depends on batch composition and is reproducible. Clamping the cap **down** is
always buffer-safe (it can only reduce the number of partial-reduce entries
relative to what the caller already allocated).

Validation (same reproducer, `AITER_MLA_DECODE_DETERMINISTIC=1`):

```
[det] split=1 vs split=8 max|d| = 0.000e+00   (clamp collapses split=8 onto split=1)
[det] PASS: deterministic single-split path.
```

Trade-off: forcing a single split removes decode split-K parallelism, so long
context / low batch decode loses some latency headroom. It is therefore **opt-in**
and intended for serving that requires reproducible outputs. The durable fix is
an **order-stable reduce** in the kernel (fixed split count + deterministic
accumulation order), which would keep both determinism and split-K performance;
happy to help test such a kernel change.

## Test plan

- [x] `op_tests/test_mla_qh16_decode_determinism.py` on MI300X (gfx942): baseline
      shows split-count-dependent numerics (`2.44e-3`) with each fixed split count
      bit-stable; `AITER_MLA_DECODE_DETERMINISTIC=1` collapses split=8 onto split=1
      (`max|d|=0`) and PASSES.
- [ ] Broader shapes (varying nhead/ctx/batch) and a perf note on the single-split
      path at long context — reviewer/maintainer guidance welcome.

Env: ROCm 7.2.3, MI300X (gfx942), AITER v0.1.18, PyTorch 2.11, GLM-5.1-FP8.

### CODE REVIEW

File: aiter/ops/attention.py [hunk 1]
```diff
@@ -1402,6 +1403,38 @@ def get_mla_metadata_v1(
...
+    if _det_cap is None and os.getenv("AITER_MLA_DECODE_DETERMINISTIC", "0") not in (
+        "0",
+        "",
+    ):
+        _det_cap = "1"
+    if _det_cap is not None:
+        try:
+            _det_cap_i = int(_det_cap)
+        except ValueError:
+            _det_cap_i = 1
```
[the remark this answers is not in the document]
Comment (danila-permogorskii) [reply aiter/ops/attention.py:1430]: Falling back to 1 seems unsafe because it converts a configuration error into a valid but potentially expensive mode. Could we ignore an invalid value instead or make a warning?

File: aiter/ops/attention.py [hunk 2]
```diff
@@ -1400,6 +1403,38 @@ def get_mla_metadata_v1(
...
+    # logits (observed as +/-1-2 NIAH needles at mid context). Forcing a single
+    # split per request makes the reduce a no-op, so the result is independent of
+    # batch composition and reproducible run-to-run. Opt-in (costs decode
+    # parallelism at long context); the durable fix is an order-stable reduce.
+    # AITER_MLA_DECODE_MAX_SPLIT_PER_BATCH=<n> pins the cap to n (n>=1);
+    # AITER_MLA_DECODE_DETERMINISTIC=1 is shorthand for n=1. Clamping down is
+    # always buffer-safe (it can only reduce the number of partials vs the size
+    # the caller allocated).
+    _det_cap = os.getenv("AITER_MLA_DECODE_MAX_SPLIT_PER_BATCH")
+    if _det_cap is None and os.getenv("AITER_MLA_DECODE_DETERMINISTIC", "0") not in (
```
Comment (danila-permogorskii) [inline aiter/ops/attention.py:1421]: Could we define what happens when both environment variables are set?

**Review by danila-permogorskii — CHANGES_REQUESTED** [review]
Hey! I think some moment should be slightly changed avoiding unexpected behaviour. I'd like to suggest make some changes in tests (valid, invalid, zero, negative and combine environment settings).

Also, MAX_SPLIT_PER_BATCH=n appears to cap rather than pin the actual split count. So the comment should avoid implying reproducibility for n > 1

File: aiter/ops/attention.py (hunk 2 above)
Comment (MohitAMD) [reply aiter/ops/attention.py]: Handled in newer commit.

File: op_tests/test_mla_qh16_decode_determinism.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+    print(f"[det] gfx={gfx} nhead={NHEAD} fp8/fp8 page_size={PAGE_SIZE} "
+          f"batch={BATCH} ctx={CTX}")
+    torch.manual_seed(0)
+    dev = "cuda"
+    q = torch.randn(BATCH * DECODE_QLEN, NHEAD, QK_HEAD_DIM, device=dev).to(dtypes.fp8)
+    kv = torch.randn(CTX, PAGE_SIZE, NHEAD_KV, QK_HEAD_DIM, device=dev).to(dtypes.fp8)
+
+    N = 8
+    det_env = os.getenv("AITER_MLA_DECODE_DETERMINISTIC", "0") not in ("0", "")
+
```
[the remark this answers is not in the document]
Comment (MohitAMD) [reply op_tests/test_mla_qh16_decode_determinism.py:167]: Handled in newer commit.

File: aiter/ops/attention.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (MohitAMD) [reply aiter/ops/attention.py]: Handled in newer commit.

**Comment by MohitAMD** [comment]
Hi @danila-permogorskii , 
I have taken your feedback and resolved the open comments in the latest commits. CI tests are passing. Please merge this PR.

**Comment by MohitAMD** [comment]
@ROCm/team_aiter , is there anything else needed? Can we expedite this PR review?

**Comment by danila-permogorskii** [comment]
Hi, Mohit!

Sorry for the late answer, I'll check this this week :-)
